In [1]:
import pandas as pd
import plotly.express as px

In [2]:
df = pd.read_csv('brasileirao_2012_a_2023.csv')
df.head()

Ano  Colocação               Clube  Pontos  Vitórias  Empates  Derrotas  \
0  2012          1     Fluminense - RJ      77        22       11         5   
1  2012          2       Atlético - MG      72        20       12         6   
2  2012          3         Grêmio - RS      71        20       11         7   
3  2012          4      São Paulo - SP      66        20        6        12   
4  2012          5  Vasco da Gama - RJ      58        16       10        12   

   Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
0           61             33              28                85   
1           64             37              27               104   
2           56             33              23               102   
3           59             37              22                93   
4           45             44               1                91   

   Cartões Vermelhos  Aproveitamento (%)  
0                  3                  67  
1                 10                  63  
2                  9                  62  
3                  6                  57  
4                  3                  50

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Ano                 240 non-null    int64 
 1   Colocação           240 non-null    int64 
 2   Clube               240 non-null    object
 3   Pontos              240 non-null    int64 
 4   Vitórias            240 non-null    int64 
 5   Empates             240 non-null    int64 
 6   Derrotas            240 non-null    int64 
 7   Gols Feitos         240 non-null    int64 
 8   Gols Sofridos       240 non-null    int64 
 9   Saldos de Gols      240 non-null    int64 
 10  Cartões Amarelos    240 non-null    int64 
 11  Cartões Vermelhos   240 non-null    int64 
 12  Aproveitamento (%)  240 non-null    int64 
dtypes: int64(12), object(1)
memory usage: 24.5+ KB


<h2>Normalizando clubes</h2>
<p>Alguns clubes aparecem com nomes distintos, a depender do ano. A tarefa aqui é identificar esses times e agrupá-los devidamente.</p>

In [4]:
df.groupby('Clube').sum()

Ano  Colocação  Pontos  Vitórias  Empates  \
Clube                                                                    
America - MG                2021          8      53        13       14   
America Fc - MG             2018         17      40        10       10   
América - MG                2016         20      28         7        7   
América Fc Saf - MG         2022         10      53        15        8   
América Saf - MG            2023         20      24         5        9   
Athletico Paranaense - PR  12123         49     335        92       59   
Atletico - PR               2014          8      54        15        9   
Atlético - GO              10092         79     205        49       58   
Atlético - MG              18144         52     551       154       89   
Atlético - PR               6048         27     159        45       24   
Atlético Mineiro - MG       4043          8     142        41       19   
Atlético Mineiro Saf - MG   2023          3      66        19        9   
Atlético Paranaense - PR    2013          3      64        18       10   
Avaí - SC                   8073         74     140        33       41   
Bahia - BA                 18157        127     410       104       98   
Botafogo - RJ              20174        105     500       136       92   
Ceará - CE                 10100         70     222        52       66   
Chapecoense - SC           14120        101     287        70       77   
Corinthians - SP           24210         84     707       188      143   
Coritiba - PR              16129        119     349        90       79   
Coritiba S.a.f. - PR        2023         19      30         8        6   
Criciuma - SC               4027         34      78        20       18   
Cruzeiro - MG              16124         61     460       127       79   
Cruzeiro Saf - MG           2023         14      47        11       14   
Csa - AL                    2019         18      32         8        8   
Cuiabá - MT                 2021         15      47        10       17   
Cuiabá Saf - MT             4045         28      92        24       20   
Figueirense - SC            8057         67     157        39       40   
Flamengo - RJ              24210         73     755       216      111   
Fluminense - RJ            24210        110     663       183      114   
Fortaleza - CE             10105         47     261        72       45   
Goiás - GO                 14126         96     317        83       68   
Grêmio - RS                22188         61     681       191      108   
Internacional - RS         22193         83     644       175      119   
Joinville - SC              2015         20      31         7       10   
Juventude - RS              4043         36      68        14       26   
Náutico - PE                4025         32      69        19       12   
Palmeiras - SP             22197         62     699       200       99   
Paraná - PR                 2018         20      23         4       11   
Ponte Preta - SP           10073         71     228        59       51   
Portuguesa - SP             4025         33      89        22       27   
Red Bull Bragantino - SP    8086         36     215        55       50   
Santa Cruz - PE             2016         19      31         8        7   
Santos - SP                24210         95     673       182      127   
Sport - PE                 16133        115     363        96       78   
São Paulo - SP             24210         90     697       188      133   
Vasco da Gama - RJ         14114         93     332        84       80   
Vasco da Gama S.a.f. - RJ   2023         15      45        12        9   
Vitória - BA               10078         73     222        58       48   

                           Derrotas  Gols Feitos  Gols Sofridos  \
Clube                                                             
America - MG                     11           41             37   
America Fc - MG                  18           30     

In [5]:
def agrupar_nomes(lista_com_nomes: list, nome_correto: str):
    for nome in lista_com_nomes:
        df.loc[df['Clube'] == nome, 'Clube'] = nome_correto

nomes_incorretos = {
    "América - MG": ["America - MG", "America Fc - MG", "América Fc Saf - MG", "América Saf - MG"],
    "Athletico Paranaense - PR": ["Atletico - PR", "Atlético - PR", "Atlético Paranaense - PR"],
    "Atlético Mineiro - MG": ["Atlético - MG", "Atlético Mineiro Saf - MG"],
    "Atlético Goianiense - GO": ["Atlético - GO"],
    "Coritiba - PR": ["Coritiba S.a.f. - PR"],
    "Cruzeiro - MG": ["Cruzeiro Saf - MG"],
    "Vasco da Gama - RJ": ["Vasco da Gama S.a.f. - RJ"],
    "Cuiabá - MT": ["Cuiabá Saf - MT"],
}

for nome_correto in nomes_incorretos.keys():
    agrupar_nomes(nomes_incorretos[nome_correto], nome_correto)

In [6]:
sorted(df['Clube'].unique())

['América - MG',
 'Athletico Paranaense - PR',
 'Atlético Goianiense - GO',
 'Atlético Mineiro - MG',
 'Avaí - SC',
 'Bahia - BA',
 'Botafogo - RJ',
 'Ceará - CE',
 'Chapecoense - SC',
 'Corinthians - SP',
 'Coritiba - PR',
 'Criciuma - SC',
 'Cruzeiro - MG',
 'Csa - AL',
 'Cuiabá - MT',
 'Figueirense - SC',
 'Flamengo - RJ',
 'Fluminense - RJ',
 'Fortaleza - CE',
 'Goiás - GO',
 'Grêmio - RS',
 'Internacional - RS',
 'Joinville - SC',
 'Juventude - RS',
 'Náutico - PE',
 'Palmeiras - SP',
 'Paraná - PR',
 'Ponte Preta - SP',
 'Portuguesa - SP',
 'Red Bull Bragantino - SP',
 'Santa Cruz - PE',
 'Santos - SP',
 'Sport - PE',
 'São Paulo - SP',
 'Vasco da Gama - RJ',
 'Vitória - BA']

In [7]:
len(df['Clube'].unique())

36

<h2>1 - Total da soma de cada coluna de cada time.</h2>

In [8]:
soma_colunas_por_time = df.groupby('Clube').sum().drop(columns=['Ano', 'Aproveitamento (%)', 'Colocação']).copy()
soma_colunas_por_time

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                         
América - MG                  198        50       48        92          176   
Athletico Paranaense - PR     612       170      102       146          517   
Atlético Goianiense - GO      205        49       58        83          187   
Atlético Mineiro - MG         759       214      117       125          670   
Avaí - SC                     140        33       41        78          119   
Bahia - BA                    410       104       98       140          378   
Botafogo - RJ                 500       136       92       152          434   
Ceará - CE                    222        52       66        72          195   
Chapecoense - SC              287        70       77       119          261   
Corinthians - SP              707       188      143       125          548   
Coritiba - PR                 379        98       85       159          362   
Criciuma - SC                  78        20       18        38           77   
Cruzeiro - MG                 507       138       93       111          426   
Csa - AL                       32         8        8        22           24   
Cuiabá - MT                   139        34       37        43          105   
Figueirense - SC              157        39       40        73          142   
Flamengo - RJ                 755       216      111       129          672   
Fluminense - RJ               663       183      114       159          577   
Fortaleza - CE                261        72       45        73          219   
Goiás - GO                    317        83       68       115          288   
Grêmio - RS                   681       191      108       119          554   
Internacional - RS            644       175      119       124          526   
Joinville - SC                 31         7       10        21           26   
Juventude - RS                 68        14       26        36           65   
Náutico - PE                   69        19       12        45           66   
Palmeiras - SP                699       200       99       119          620   
Paraná - PR                    23         4       11        23           18   
Ponte Preta - SP              228        59       51        80          200   
Portuguesa - SP                89        22       27        27           89   
Red Bull Bragantino - SP      215        55       50        47          203   
Santa Cruz - PE                31         8        7        23           45   
Santos - SP                   673       182      127       147          579   
Sport - PE                    363        96       78       130          313   
São Paulo - SP                697       188      133       135          572   
Vasco da Gama - RJ            377        96       89       119          321   
Vitória - BA                  222        58       48        84          233   

                           Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                        
América - MG                         263             -87               390   
Athletico Paranaense - PR            455              62               953   
Atlético Goianiense - GO             261             -74               469   
Atlético Mineiro - MG                502             168              1066   
Avaí - SC                            230            -111               382   
Bahia - BA                           424             -46               826   
Botafogo - RJ                        453             -19               880   
Ceará - CE                           209             -14               484   
Chapecoense - SC                     362            -101               566   
Corinthians - SP                     429             119               916   
Coritiba - PR                        472            -110               873   
Criciuma -

<h2>2 - Média de cada coluna de cada time.</h2>

In [9]:
media_colunas_por_time = df.groupby('Clube').mean().drop(columns=['Ano']).round(1).copy()
media_colunas_por_time

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                       
América - MG                    15.0    39.6      10.0      9.6      18.4   
Athletico Paranaense - PR        7.9    55.6      15.5      9.3      13.3   
Atlético Goianiense - GO        15.8    41.0       9.8     11.6      16.6   
Atlético Mineiro - MG            5.2    63.2      17.8      9.8      10.4   
Avaí - SC                       18.5    35.0       8.2     10.2      19.5   
Bahia - BA                      14.1    45.6      11.6     10.9      15.6   
Botafogo - RJ                   10.5    50.0      13.6      9.2      15.2   
Ceará - CE                      14.0    44.4      10.4     13.2      14.4   
Chapecoense - SC                14.4    41.0      10.0     11.0      17.0   
Corinthians - SP                 7.0    58.9      15.7     11.9      10.4   
Coritiba - PR                   15.3    42.1      10.9      9.4      17.7   
Criciuma - SC                   17.0    39.0      10.0      9.0      19.0   
Cruzeiro - MG                    8.3    56.3      15.3     10.3      12.3   
Csa - AL                        18.0    32.0       8.0      8.0      22.0   
Cuiabá - MT                     14.3    46.3      11.3     12.3      14.3   
Figueirense - SC                16.8    39.2       9.8     10.0      18.2   
Flamengo - RJ                    6.1    62.9      18.0      9.2      10.8   
Fluminense - RJ                  9.2    55.2      15.2      9.5      13.2   
Fortaleza - CE                   9.4    52.2      14.4      9.0      14.6   
Goiás - GO                      13.7    45.3      11.9      9.7      16.4   
Grêmio - RS                      5.5    61.9      17.4      9.8      10.8   
Internacional - RS               7.5    58.5      15.9     10.8      11.3   
Joinville - SC                  20.0    31.0       7.0     10.0      21.0   
Juventude - RS                  18.0    34.0       7.0     13.0      18.0   
Náutico - PE                    16.0    34.5       9.5      6.0      22.5   
Palmeiras - SP                   5.6    63.5      18.2      9.0      10.8   
Paraná - PR                     20.0    23.0       4.0     11.0      23.0   
Ponte Preta - SP                14.2    45.6      11.8     10.2      16.0   
Portuguesa - SP                 16.5    44.5      11.0     13.5      13.5   
Red Bull Bragantino - SP         9.0    53.8      13.8     12.5      11.8   
Santa Cruz - PE                 19.0    31.0       8.0      7.0      23.0   
Santos - SP                      7.9    56.1      15.2     10.6      12.2   
Sport - PE                      14.4    45.4      12.0      9.8      16.2   
São Paulo - SP                   7.5    58.1      15.7     11.1      11.2   
Vasco da Gama - RJ              13.5    47.1      12.0     11.1      14.9   
Vitória - BA                    14.6    44.4      11.6      9.6      16.8   

                           Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                                   
América - MG                      35.2           52.6           -17.4   
Athletico Paranaense - PR         47.0           41.4             5.6   
Atlético Goianiense - GO          37.4           52.2           -14.8   
Atlético Mineiro - MG             55.8           41.8            14.0   
Avaí - SC                         29.8           57.5           -27.8   
Bahia - BA                        42.0           47.1            -5.1   
Botafogo - RJ                     43.4           45.3            -1.9   
Ceará - CE                        39.0           41.8            -2.8   
Chapecoense - SC                  37.3           51.7           -14.4   
Corinthians - SP                  45.7           35.8             9.9   
Coritiba - PR                     40.2           52.4           -12.2   
Criciuma - SC                     38.5           59.5           -21.0   
Cruzeiro - MG                     47.3           40.1             7.2   
Csa -

<h2>3 - Soma das informações de cada ano.</h2>

In [10]:
somas_colunas_por_ano = df.groupby('Ano').sum().drop(columns=['Clube', 'Aproveitamento (%)', 'Colocação']).copy()
somas_colunas_por_ano

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Ano                                                                     
2012    1035       275      210       275          940            940   
2013    1024       272      216       272          936            936   
2014    1048       288      184       288          860            860   
2015    1049       289      182       289          897            897   
2016    1043       285      188       287          912            918   
2017    1037       277      206       277          923            923   
2018    1027       270      220       270          827            827   
2019    1042       282      196       282          875            875   
2020    1032       272      216       272          944            944   
2021    1027       267      226       267          842            842   
2022    1032       272      216       272          905            905   
2023    1042       282      196       282          946            946   

      Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  
Ano                                                        
2012               0              1893                112  
2013               0              1683                107  
2014               0              1657                 82  
2015               0              1824                109  
2016              -6              1693                 86  
2017               0              1777                 77  
2018               0              1815                101  
2019               0              1703                116  
2020               0              1730                133  
2021               0              1803                116  
2022               0              1948                143  
2023               0              2033                134

<p>Em 2016, na última rodada do Brasileirão, houve um W.O duplo entre Chapecoense x Atlético-MG. Devido à isso, o saldo de gols é negativo.</p>

<h2>4 - Média de cada informação de cada ano</h2>

In [11]:
media_colunas_por_anos = df.groupby('Ano').mean(numeric_only=True).drop(columns="Colocação").round(1).copy()
media_colunas_por_anos

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Ano                                                                     
2012    51.8      13.8     10.5      13.8         47.0           47.0   
2013    51.2      13.6     10.8      13.6         46.8           46.8   
2014    52.4      14.4      9.2      14.4         43.0           43.0   
2015    52.4      14.4      9.1      14.4         44.8           44.8   
2016    52.2      14.2      9.4      14.4         45.6           45.9   
2017    51.8      13.8     10.3      13.8         46.2           46.2   
2018    51.4      13.5     11.0      13.5         41.4           41.4   
2019    52.1      14.1      9.8      14.1         43.8           43.8   
2020    51.6      13.6     10.8      13.6         47.2           47.2   
2021    51.4      13.4     11.3      13.4         42.1           42.1   
2022    51.6      13.6     10.8      13.6         45.2           45.2   
2023    52.1      14.1      9.8      14.1         47.3           47.3   

      Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Ano                                                                            
2012             0.0              94.6                5.6                45.0  
2013             0.0              84.2                5.4                44.5  
2014             0.0              82.8                4.1                45.6  
2015             0.0              91.2                5.4                45.4  
2016            -0.3              84.6                4.3                45.4  
2017             0.0              88.8                3.8                45.0  
2018             0.0              90.8                5.0                44.6  
2019             0.0              85.2                5.8                45.2  
2020             0.0              86.5                6.6                44.6  
2021             0.0              90.2                5.8                44.6  
2022             0.0              97.4                7.2                44.8  
2023             0.0             101.6                6.7                45.2

<h2>5 - Identificar A MENOR e A MAIOR pontuação do 17 colocado no Brasileirão.</h2>

In [12]:
def devolve_maior_e_menor_pontuacao(colocacao: int | float):
    tabela = df.loc[df['Colocação'] == colocacao, :].copy()
    if tabela.empty:     
        return None
    tabela = tabela.sort_values('Pontos', ascending=False)
    return {"maior": tabela.head(1), "menor": tabela.tail(1)}

In [13]:
colocacoes_17 = devolve_maior_e_menor_pontuacao(17)
colocacoes_17['maior']

Ano  Colocação            Clube  Pontos  Vitórias  Empates  Derrotas  \
36  2013         17  Portuguesa - SP      44        12       12        14   

    Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
36           50             46               4                89   

    Cartões Vermelhos  Aproveitamento (%)  
36                  9                  38

In [14]:
colocacoes_17['menor']

Ano  Colocação          Clube  Pontos  Vitórias  Empates  Derrotas  \
156  2019         17  Cruzeiro - MG      36         7       15        16   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
156           27             46             -19                81   

     Cartões Vermelhos  Aproveitamento (%)  
156                  6                  31

<h2>5.1 - Média de pontuação do 17 colocado.</h2>

In [81]:
df.loc[df['Colocação'] == 17, 'Pontos'].mean()

40.916666666666664

<h2>6 - Identificar A MENOR e A MAIOR pontuação do lanterna do Brasileirão.</h2>

In [15]:
colocacoes_20 = devolve_maior_e_menor_pontuacao(20)
colocacoes_20['maior']

Ano  Colocação                     Clube  Pontos  Vitórias  Empates  \
119  2017         20  Atlético Goianiense - GO      36         9        9   

     Derrotas  Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
119        20           38             56             -18                83   

     Cartões Vermelhos  Aproveitamento (%)  
119                  3                  31

In [16]:
colocacoes_20['menor']

Ano  Colocação             Clube  Pontos  Vitórias  Empates  Derrotas  \
199  2021         20  Chapecoense - SC      15         1       12        25   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
199           27             67             -40                82   

     Cartões Vermelhos  Aproveitamento (%)  
199                 10                  13

<h2>6.1 - Média de pontuação do lanterna.</h2>

In [82]:
df.loc[df['Colocação'] == 20, 'Pontos'].mean()

25.666666666666668

<h2>7 - Identificar A MENOR e A MAIOR pontuação do campeão no Brasileirão.</h2>

In [17]:
colocacoes_1 = devolve_maior_e_menor_pontuacao(1)
colocacoes_1['maior']

Ano  Colocação          Clube  Pontos  Vitórias  Empates  Derrotas  \
140  2019          1  Flamengo - RJ      90        28        6         4   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
140           86             37              49                90   

     Cartões Vermelhos  Aproveitamento (%)  
140                  3                  78

In [18]:
colocacoes_1['menor']

Ano  Colocação           Clube  Pontos  Vitórias  Empates  Derrotas  \
220  2023          1  Palmeiras - SP      70        20       10         8   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
220           64             33              31               101   

     Cartões Vermelhos  Aproveitamento (%)  
220                  7                  61

<h2>7.1 - Média de pontuação do campeão.</h2>

In [83]:
df.loc[df['Colocação'] == 1, 'Pontos'].mean()

78.5

<h2>8 - Identificar o time que MAIS TOMOU CARTÃO AMARELO e o que MAIS TOMOU CARTÃO VERMELHO em UM ANO, e vice-versa.</h2>

In [19]:
#time que mais tomou cartões amarelos em uma edição
df.loc[df['Cartões Amarelos'] == df['Cartões Amarelos'].max(), :]

Ano  Colocação            Clube  Pontos  Vitórias  Empates  Derrotas  \
226  2023          7  Fluminense - RJ      56        16        8        14   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
226           51             47               4               130   

     Cartões Vermelhos  Aproveitamento (%)  
226                 12                  49

In [20]:
#time que menos tomou cartões amarelos em uma edição
df.loc[df['Cartões Amarelos'] == df['Cartões Amarelos'].min(), :]

Ano  Colocação             Clube  Pontos  Vitórias  Empates  Derrotas  \
184  2021          5  Corinthians - SP      57        15       12        11   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
184           40             36               4                61   

     Cartões Vermelhos  Aproveitamento (%)  
184                  3                  50

In [21]:
#time que mais tomou cartões vermelhos em uma edição
df.loc[df['Cartões Vermelhos'] == df['Cartões Vermelhos'].max(), :]

Ano  Colocação               Clube  Pontos  Vitórias  Empates  Derrotas  \
77  2015         18  Vasco da Gama - RJ      41        10       11        17   

    Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
77           28             54             -26               108   

    Cartões Vermelhos  Aproveitamento (%)  
77                 14                  35

In [22]:
#time que menos tomou cartões vermelhos em uma edição
df.loc[df['Cartões Vermelhos'] == df['Cartões Vermelhos'].min(), :]

Ano  Colocação           Clube  Pontos  Vitórias  Empates  Derrotas  \
80  2016          1  Palmeiras - SP      80        24        8         6   

    Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
80           62             32              30                95   

    Cartões Vermelhos  Aproveitamento (%)  
80                  0                  70

<h2>8.1 O time que mais tomou cartões amarelos e o que mais tomou vermelhos em todo o registro</h2>

In [23]:
#Mais sofreu cartões amarelos
soma_colunas_por_time.loc[soma_colunas_por_time['Cartões Amarelos'] == soma_colunas_por_time['Cartões Amarelos'].max()]

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                              
São Paulo - SP     697       188      133       135          572   

                Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                             
São Paulo - SP            473              99              1092   

                Cartões Vermelhos  
Clube                              
São Paulo - SP                 72

In [24]:
#Mais sofreu cartões vermelhos
soma_colunas_por_time.loc[soma_colunas_por_time['Cartões Vermelhos'] == soma_colunas_por_time['Cartões Vermelhos'].max()]

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                               
Fluminense - RJ     663       183      114       159          577   

                 Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                              
Fluminense - RJ            529              48              1085   

                 Cartões Vermelhos  
Clube                               
Fluminense - RJ                 82

<h2>9 - Identificar os times com as maiores e menores médias de cartões amarelos e cartões vermelhos por campeonato.</h2>
<p>Estabelecido um filtro mínimo de 3 campeonatos, isto é, apenas equipes que jogaram pelo menos 3 vezes o Brasileirão, nesse registro, serão contabilizados no cálculo.</p>

In [25]:
filtro_minimo = soma_colunas_por_time[soma_colunas_por_time['Vitórias'] + soma_colunas_por_time['Empates'] + soma_colunas_por_time['Derrotas'] <= 76].index
filtro_minimo

Index(['Criciuma - SC', 'Csa - AL', 'Joinville - SC', 'Juventude - RS',
       'Náutico - PE', 'Paraná - PR', 'Portuguesa - SP', 'Santa Cruz - PE'],
      dtype='object', name='Clube')

In [26]:
media_com_filtro = media_colunas_por_time.drop(filtro_minimo)

In [27]:
#clube com a maior média de cartões amarelos
media_com_filtro[media_com_filtro['Cartões Amarelos'] == media_com_filtro['Cartões Amarelos'].max()]

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                
Internacional - RS        7.5    58.5      15.9     10.8      11.3   

                    Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                            
Internacional - RS         47.8           39.3             8.5   

                    Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                        
Internacional - RS              99.2                6.4                51.0

In [28]:
#clube com a menor média de cartões amarelos
media_com_filtro[media_com_filtro['Cartões Amarelos'] == media_com_filtro['Cartões Amarelos'].min()]

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                           
Corinthians - SP        7.0    58.9      15.7     11.9      10.4         45.7   

                  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                               
Corinthians - SP           35.8             9.9              76.3   

                  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                    
Corinthians - SP                3.7                51.3

In [29]:
#clube com a maior média de cartões amarelos
media_com_filtro[media_com_filtro['Cartões Vermelhos'] == media_com_filtro['Cartões Vermelhos'].max()]

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                         
Fortaleza - CE        9.4    52.2      14.4      9.0      14.6         43.8   

                Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                             
Fortaleza - CE           44.2            -0.4              90.0   

                Cartões Vermelhos  Aproveitamento (%)  
Clube                                                  
Fortaleza - CE                8.2                45.2

In [30]:
#clube com a menor média de cartões amarelos
media_com_filtro[media_com_filtro['Cartões Vermelhos'] == media_com_filtro['Cartões Vermelhos'].min()]

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                           
Corinthians - SP        7.0    58.9      15.7     11.9      10.4         45.7   

                  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                               
Corinthians - SP           35.8             9.9              76.3   

                  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                    
Corinthians - SP                3.7                51.3

In [31]:
def plotar_grafico_de_cartoes(dataframe: pd.DataFrame, coluna: str, tipo: str):
    title = f"Média de {coluna} por Clube"
    if tipo == "Ano":
        title += " por Ano"
    fig = px.bar(dataframe, x=dataframe.index, y=coluna, title=title, color=dataframe.index)
    fig.update_layout(
        showlegend=False,
        xaxis_title='',
        yaxis_title=f"Média de {coluna}"
        )
    fig.update_traces(
        hovertemplate="%{x}<br>%{y}<extra></extra>"
    )
    fig.show()

In [32]:
plotar_grafico_de_cartoes(media_com_filtro, "Cartões Amarelos", "Clube")

In [33]:
plotar_grafico_de_cartoes(media_com_filtro, "Cartões Vermelhos", "Clube")

In [34]:
plotar_grafico_de_cartoes(media_colunas_por_anos, "Cartões Amarelos", "Ano")

In [35]:
plotar_grafico_de_cartoes(media_colunas_por_anos, "Cartões Vermelhos", "Ano")

<h2>10 - Identificar o time com MAIS GOLS FEITOS em uma edição.</h2>

In [36]:
df.loc[df['Gols Feitos'] == df['Gols Feitos'].max(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Gols Feitos']]

Ano  Colocação          Clube  Pontos  Gols Feitos
140  2019          1  Flamengo - RJ      90           86

<h2>10.1 - Identificar o time com MENOS GOLS FEITOS em uma edição.</h2>

In [37]:
df.loc[df['Gols Feitos'] == df['Gols Feitos'].min(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Gols Feitos']]

Ano  Colocação        Clube  Pontos  Gols Feitos
139  2018         20  Paraná - PR      23           18
159  2019         20    Avaí - SC      20           18

<h2>11 - Identificar o time com MAIS GOLS SOFRIDOS em uma edição</h2>

In [38]:
df.loc[df['Gols Sofridos'] == df['Gols Sofridos'].max(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Gols Sofridos']]

Ano  Colocação         Clube  Pontos  Gols Sofridos
239  2023         20  América - MG      24             81

<h2>11.1 - Identificar o time com MENOS GOLS SOFRIDOS em uma edição</h2>

In [39]:
df.loc[df['Gols Sofridos'] == df['Gols Sofridos'].min(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Gols Sofridos']]

Ano  Colocação             Clube  Pontos  Gols Sofridos
29  2013         10  Corinthians - SP      50             22

<h2>12 - Identificar o time com o MAIOR SALDO DE GOLS em uma edição</h2>

In [40]:
df.loc[df['Saldos de Gols'] == df['Saldos de Gols'].max(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Saldos de Gols']]

Ano  Colocação          Clube  Pontos  Saldos de Gols
140  2019          1  Flamengo - RJ      90              49

<h2>12.1 - Identificar o time com o MENOR SALDO DE GOLS em uma edição</h2>

In [41]:
df.loc[df['Saldos de Gols'] == df['Saldos de Gols'].min(), ['Ano', 'Colocação', 'Clube', 'Pontos', 'Saldos de Gols']]

Ano  Colocação         Clube  Pontos  Saldos de Gols
39  2013         20  Náutico - PE      20             -57

<h2>13 - Identificar OS TIMES que MAIS FIZERAM GOLS no TOTAL.</h2>

In [42]:
soma_colunas_por_time.sort_values('Gols Feitos', ascending=False).head()

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                     
Flamengo - RJ             755       216      111       129          672   
Atlético Mineiro - MG     759       214      117       125          670   
Palmeiras - SP            699       200       99       119          620   
Santos - SP               673       182      127       147          579   
Fluminense - RJ           663       183      114       159          577   

                       Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                    
Flamengo - RJ                    496             176               950   
Atlético Mineiro - MG            502             168              1066   
Palmeiras - SP                   439             181              1035   
Santos - SP                      494              85              1072   
Fluminense - RJ                  529              48              1085   

                       Cartões Vermelhos  
Clube                                     
Flamengo - RJ                         58  
Atlético Mineiro - MG                 60  
Palmeiras - SP                        61  
Santos - SP                           69  
Fluminense - RJ                       82

<h2>14 - Identificar OS TIMES que MAIS SOFRERAM GOLS no TOTAL.</h2>

In [43]:
soma_colunas_por_time.sort_values('Gols Sofridos', ascending=False).head()

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                     
Fluminense - RJ           663       183      114       159          577   
Atlético Mineiro - MG     759       214      117       125          670   
Flamengo - RJ             755       216      111       129          672   
Santos - SP               673       182      127       147          579   
São Paulo - SP            697       188      133       135          572   

                       Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                    
Fluminense - RJ                  529              48              1085   
Atlético Mineiro - MG            502             168              1066   
Flamengo - RJ                    496             176               950   
Santos - SP                      494              85              1072   
São Paulo - SP                   473              99              1092   

                       Cartões Vermelhos  
Clube                                     
Fluminense - RJ                       82  
Atlético Mineiro - MG                 60  
Flamengo - RJ                         58  
Santos - SP                           69  
São Paulo - SP                        72

Nos cálculos de média a seguir, assim como nos anteriores, apenas serão consideradas equipes que jogaram, pelo menos, 3 vezes o campeonato.

<h2>15 - Identificar OS TIMES que MAIS FAZEM GOLS na MÉDIA.</h2>

In [44]:
media_com_filtro.sort_values('Gols Feitos', ascending=False).head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                      
Palmeiras - SP                  5.6    63.5      18.2      9.0      10.8   
Flamengo - RJ                   6.1    62.9      18.0      9.2      10.8   
Atlético Mineiro - MG           5.2    63.2      17.8      9.8      10.4   
Red Bull Bragantino - SP        9.0    53.8      13.8     12.5      11.8   
Grêmio - RS                     5.5    61.9      17.4      9.8      10.8   

                          Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                                  
Palmeiras - SP                   56.4           39.9            16.5   
Flamengo - RJ                    56.0           41.3            14.7   
Atlético Mineiro - MG            55.8           41.8            14.0   
Red Bull Bragantino - SP         50.8           45.0             5.8   
Grêmio - RS                      50.4           37.9            12.5   

                          Cartões Amarelos  Cartões Vermelhos  \
Clube                                                           
Palmeiras - SP                        94.1                5.5   
Flamengo - RJ                         79.2                4.8   
Atlético Mineiro - MG                 88.8                5.0   
Red Bull Bragantino - SP              94.8                5.2   
Grêmio - RS                           88.5                4.7   

                          Aproveitamento (%)  
Clube                                         
Palmeiras - SP                          55.3  
Flamengo - RJ                           54.7  
Atlético Mineiro - MG                   55.1  
Red Bull Bragantino - SP                46.8  
Grêmio - RS                             53.8

<h2>15.1 - Identificar OS TIMES que MENOS FAZEM GOLS na MÉDIA.</h2>

In [45]:
media_com_filtro.sort_values('Gols Feitos').head()

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                           
Avaí - SC              18.5    35.0       8.2     10.2      19.5         29.8   
Cuiabá - MT            14.3    46.3      11.3     12.3      14.3         35.0   
América - MG           15.0    39.6      10.0      9.6      18.4         35.2   
Figueirense - SC       16.8    39.2       9.8     10.0      18.2         35.5   
Chapecoense - SC       14.4    41.0      10.0     11.0      17.0         37.3   

                  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                               
Avaí - SC                  57.5           -27.8              95.5   
Cuiabá - MT                39.3            -4.3              89.3   
América - MG               52.6           -17.4              78.0   
Figueirense - SC           54.8           -19.2              96.0   
Chapecoense - SC           51.7           -14.4              80.9   

                  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                    
Avaí - SC                       5.5                30.0  
Cuiabá - MT                     7.3                40.0  
América - MG                    5.8                34.4  
Figueirense - SC                8.0                34.0  
Chapecoense - SC                4.0                35.7

In [46]:
fig = px.bar(media_com_filtro, x=media_com_filtro.index, y="Gols Feitos", color=media_com_filtro.index, title="Média de Gols Feitos por Equipe")
fig.update_layout(
    showlegend=False,
    yaxis_title = "Média de Gols Feitos",
    xaxis_title = "",
)
fig.show()

<h2>16 - Identificar OS TIMES que MAIS SOFREM GOLS na MÉDIA.</h2>

In [47]:
media_com_filtro.sort_values('Gols Sofridos', ascending=False).head()

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                           
Avaí - SC              18.5    35.0       8.2     10.2      19.5         29.8   
Vitória - BA           14.6    44.4      11.6      9.6      16.8         46.6   
Figueirense - SC       16.8    39.2       9.8     10.0      18.2         35.5   
América - MG           15.0    39.6      10.0      9.6      18.4         35.2   
Coritiba - PR          15.3    42.1      10.9      9.4      17.7         40.2   

                  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                               
Avaí - SC                  57.5           -27.8              95.5   
Vitória - BA               56.2            -9.6              89.8   
Figueirense - SC           54.8           -19.2              96.0   
América - MG               52.6           -17.4              78.0   
Coritiba - PR              52.4           -12.2              97.0   

                  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                    
Avaí - SC                       5.5                30.0  
Vitória - BA                    4.6                38.4  
Figueirense - SC                8.0                34.0  
América - MG                    5.8                34.4  
Coritiba - PR                   5.9                36.6

<h2>16.1 - Identificar OS TIMES que MENOS SOFREM GOLS na MÉDIA.</h2>

In [48]:
media_com_filtro.sort_values('Gols Sofridos').head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                
Corinthians - SP          7.0    58.9      15.7     11.9      10.4   
Grêmio - RS               5.5    61.9      17.4      9.8      10.8   
Internacional - RS        7.5    58.5      15.9     10.8      11.3   
Cuiabá - MT              14.3    46.3      11.3     12.3      14.3   
São Paulo - SP            7.5    58.1      15.7     11.1      11.2   

                    Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                            
Corinthians - SP           45.7           35.8             9.9   
Grêmio - RS                50.4           37.9            12.5   
Internacional - RS         47.8           39.3             8.5   
Cuiabá - MT                35.0           39.3            -4.3   
São Paulo - SP             47.7           39.4             8.2   

                    Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                        
Corinthians - SP                76.3                3.7                51.3  
Grêmio - RS                     88.5                4.7                53.8  
Internacional - RS              99.2                6.4                51.0  
Cuiabá - MT                     89.3                7.3                40.0  
São Paulo - SP                  91.0                6.0                50.4

In [49]:
fig = px.bar(media_com_filtro, x=media_com_filtro.index, y="Gols Sofridos", color=media_com_filtro.index, title="Média de Gols Sofridos por Equipe")
fig.update_layout(
    showlegend=False,
    yaxis_title = "Média de Gols Sofridos",
    xaxis_title = "",
)
fig.show()

<h2>17 - Identificar OS TIMES com MAIOR SALDO DE GOLS na MÉDIA.</h2>

In [50]:
media_com_filtro.sort_values('Saldos de Gols', ascending=False).head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                   
Palmeiras - SP               5.6    63.5      18.2      9.0      10.8   
Flamengo - RJ                6.1    62.9      18.0      9.2      10.8   
Atlético Mineiro - MG        5.2    63.2      17.8      9.8      10.4   
Grêmio - RS                  5.5    61.9      17.4      9.8      10.8   
Corinthians - SP             7.0    58.9      15.7     11.9      10.4   

                       Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                               
Palmeiras - SP                56.4           39.9            16.5   
Flamengo - RJ                 56.0           41.3            14.7   
Atlético Mineiro - MG         55.8           41.8            14.0   
Grêmio - RS                   50.4           37.9            12.5   
Corinthians - SP              45.7           35.8             9.9   

                       Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                           
Palmeiras - SP                     94.1                5.5                55.3  
Flamengo - RJ                      79.2                4.8                54.7  
Atlético Mineiro - MG              88.8                5.0                55.1  
Grêmio - RS                        88.5                4.7                53.8  
Corinthians - SP                   76.3                3.7                51.3

<h2>17.1 - Identificar OS TIMES com MENOR SALDO DE GOLS na MÉDIA.</h2>

In [51]:
media_com_filtro.sort_values('Saldos de Gols').head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                      
Avaí - SC                      18.5    35.0       8.2     10.2      19.5   
Figueirense - SC               16.8    39.2       9.8     10.0      18.2   
América - MG                   15.0    39.6      10.0      9.6      18.4   
Atlético Goianiense - GO       15.8    41.0       9.8     11.6      16.6   
Chapecoense - SC               14.4    41.0      10.0     11.0      17.0   

                          Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                                  
Avaí - SC                        29.8           57.5           -27.8   
Figueirense - SC                 35.5           54.8           -19.2   
América - MG                     35.2           52.6           -17.4   
Atlético Goianiense - GO         37.4           52.2           -14.8   
Chapecoense - SC                 37.3           51.7           -14.4   

                          Cartões Amarelos  Cartões Vermelhos  \
Clube                                                           
Avaí - SC                             95.5                5.5   
Figueirense - SC                      96.0                8.0   
América - MG                          78.0                5.8   
Atlético Goianiense - GO              93.8                6.2   
Chapecoense - SC                      80.9                4.0   

                          Aproveitamento (%)  
Clube                                         
Avaí - SC                               30.0  
Figueirense - SC                        34.0  
América - MG                            34.4  
Atlético Goianiense - GO                35.4  
Chapecoense - SC                        35.7

In [52]:
fig = px.bar(media_com_filtro, x=media_com_filtro.index, y="Saldos de Gols", color=media_com_filtro.index, title="Média de Saldo de Gols por Equipe")
fig.update_layout(
    showlegend=False,
    yaxis_title = "Média de Saldo de Gols",
    xaxis_title = "",
)
fig.show()

<h2>18 - Identificar o time c/ mais vitórias em uma edição</h2>

In [53]:
df[df['Vitórias'] == df['Vitórias'].max()]

Ano  Colocação          Clube  Pontos  Vitórias  Empates  Derrotas  \
140  2019          1  Flamengo - RJ      90        28        6         4   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
140           86             37              49                90   

     Cartões Vermelhos  Aproveitamento (%)  
140                  3                  78

<h2>19 - Identificar o time c/ mais empates em uma edição</h2>

In [54]:
df[df['Empates'] == df['Empates'].max()]

Ano  Colocação             Clube  Pontos  Vitórias  Empates  Derrotas  \
29   2013         10  Corinthians - SP      50        11       17        10   
165  2020          6       Grêmio - RS      59        14       17         7   
190  2021         11        Ceará - CE      50        11       17        10   
194  2021         15       Cuiabá - MT      47        10       17        11   

     Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
29            27             22               5                66   
165           53             40              13                70   
190           39             38               1                90   
194           34             37              -3                84   

     Cartões Vermelhos  Aproveitamento (%)  
29                   3                  43  
165                  8                  51  
190                  5                  43  
194                  7                  41

<h2>20 - Identificar o time c/ mais derrotas em uma edição</h2>

In [55]:
df[df['Derrotas'] == df['Derrotas'].max()]

Ano  Colocação         Clube  Pontos  Vitórias  Empates  Derrotas  \
39  2013         20  Náutico - PE      20         5        5        28   

    Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
39           22             79             -57                77   

    Cartões Vermelhos  Aproveitamento (%)  
39                  3                  17

<h2>21 - Identificar OS TIMES C/ MAIS VITÓRIAS no TOTAL</h1>

In [56]:
soma_colunas_por_time.sort_values('Vitórias', ascending=False).head(5)

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                     
Flamengo - RJ             755       216      111       129          672   
Atlético Mineiro - MG     759       214      117       125          670   
Palmeiras - SP            699       200       99       119          620   
Grêmio - RS               681       191      108       119          554   
São Paulo - SP            697       188      133       135          572   

                       Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                    
Flamengo - RJ                    496             176               950   
Atlético Mineiro - MG            502             168              1066   
Palmeiras - SP                   439             181              1035   
Grêmio - RS                      417             137               974   
São Paulo - SP                   473              99              1092   

                       Cartões Vermelhos  
Clube                                     
Flamengo - RJ                         58  
Atlético Mineiro - MG                 60  
Palmeiras - SP                        61  
Grêmio - RS                           52  
São Paulo - SP                        72

<h2>22 - Identificar OS TIMES C/ MAIS EMPATES no TOTAL</h2>

In [57]:
soma_colunas_por_time.sort_values('Empates', ascending=False).head(5)

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                     
Corinthians - SP          707       188      143       125          548   
São Paulo - SP            697       188      133       135          572   
Santos - SP               673       182      127       147          579   
Internacional - RS        644       175      119       124          526   
Atlético Mineiro - MG     759       214      117       125          670   

                       Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                    
Corinthians - SP                 429             119               916   
São Paulo - SP                   473              99              1092   
Santos - SP                      494              85              1072   
Internacional - RS               432              94              1091   
Atlético Mineiro - MG            502             168              1066   

                       Cartões Vermelhos  
Clube                                     
Corinthians - SP                      44  
São Paulo - SP                        72  
Santos - SP                           69  
Internacional - RS                    70  
Atlético Mineiro - MG                 60

<h2>23 - Identificar OS TIMES C/ MAIS DERROTAS no TOTAL</h2>

In [58]:
soma_colunas_por_time.sort_values('Derrotas', ascending=False).head(5)

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                         
Coritiba - PR                 379        98       85       159          362   
Fluminense - RJ               663       183      114       159          577   
Botafogo - RJ                 500       136       92       152          434   
Santos - SP                   673       182      127       147          579   
Athletico Paranaense - PR     612       170      102       146          517   

                           Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                                        
Coritiba - PR                        472            -110               873   
Fluminense - RJ                      529              48              1085   
Botafogo - RJ                        453             -19               880   
Santos - SP                          494              85              1072   
Athletico Paranaense - PR            455              62               953   

                           Cartões Vermelhos  
Clube                                         
Coritiba - PR                             53  
Fluminense - RJ                           82  
Botafogo - RJ                             47  
Santos - SP                               69  
Athletico Paranaense - PR                 43

<h2>24 - Identificar OS TIMES C/ MAIS VITÓRIAS na MÉDIA</h2>

In [59]:
media_com_filtro.sort_values('Vitórias', ascending=False).head(5)

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                   
Palmeiras - SP               5.6    63.5      18.2      9.0      10.8   
Flamengo - RJ                6.1    62.9      18.0      9.2      10.8   
Atlético Mineiro - MG        5.2    63.2      17.8      9.8      10.4   
Grêmio - RS                  5.5    61.9      17.4      9.8      10.8   
Internacional - RS           7.5    58.5      15.9     10.8      11.3   

                       Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                               
Palmeiras - SP                56.4           39.9            16.5   
Flamengo - RJ                 56.0           41.3            14.7   
Atlético Mineiro - MG         55.8           41.8            14.0   
Grêmio - RS                   50.4           37.9            12.5   
Internacional - RS            47.8           39.3             8.5   

                       Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                           
Palmeiras - SP                     94.1                5.5                55.3  
Flamengo - RJ                      79.2                4.8                54.7  
Atlético Mineiro - MG              88.8                5.0                55.1  
Grêmio - RS                        88.5                4.7                53.8  
Internacional - RS                 99.2                6.4                51.0

In [60]:
def media_de_equipe_da_estatistica(coluna):
    fig = px.bar(media_com_filtro, x=media_com_filtro.index, y=coluna, color=media_com_filtro.index, title=f"Média de {coluna} por Clube")
    fig.update_layout(
    showlegend=False,
    xaxis_title = "",
    yaxis_title = f"Média de {coluna}"
    )
    fig.show()

In [61]:
media_de_equipe_da_estatistica("Vitórias")

<h2>25 - Identificar OS TIMES C/ MAIS EMPATES na MÉDIA</h2>

In [62]:
media_com_filtro.sort_values('Empates', ascending=False).head(5)

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                      
Ceará - CE                     14.0    44.4      10.4     13.2      14.4   
Red Bull Bragantino - SP        9.0    53.8      13.8     12.5      11.8   
Cuiabá - MT                    14.3    46.3      11.3     12.3      14.3   
Corinthians - SP                7.0    58.9      15.7     11.9      10.4   
Atlético Goianiense - GO       15.8    41.0       9.8     11.6      16.6   

                          Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                                  
Ceará - CE                       39.0           41.8            -2.8   
Red Bull Bragantino - SP         50.8           45.0             5.8   
Cuiabá - MT                      35.0           39.3            -4.3   
Corinthians - SP                 45.7           35.8             9.9   
Atlético Goianiense - GO         37.4           52.2           -14.8   

                          Cartões Amarelos  Cartões Vermelhos  \
Clube                                                           
Ceará - CE                            96.8                6.2   
Red Bull Bragantino - SP              94.8                5.2   
Cuiabá - MT                           89.3                7.3   
Corinthians - SP                      76.3                3.7   
Atlético Goianiense - GO              93.8                6.2   

                          Aproveitamento (%)  
Clube                                         
Ceará - CE                              38.4  
Red Bull Bragantino - SP                46.8  
Cuiabá - MT                             40.0  
Corinthians - SP                        51.3  
Atlético Goianiense - GO                35.4

In [63]:
media_de_equipe_da_estatistica("Empates")

<h2>26 - Identificar OS TIMES C/ MAIS DERROTAS na MÉDIA</h2>

In [64]:
media_com_filtro.sort_values('Derrotas', ascending=False).head(5)

Colocação  Pontos  Vitórias  Empates  Derrotas  Gols Feitos  \
Clube                                                                           
Avaí - SC              18.5    35.0       8.2     10.2      19.5         29.8   
América - MG           15.0    39.6      10.0      9.6      18.4         35.2   
Figueirense - SC       16.8    39.2       9.8     10.0      18.2         35.5   
Coritiba - PR          15.3    42.1      10.9      9.4      17.7         40.2   
Chapecoense - SC       14.4    41.0      10.0     11.0      17.0         37.3   

                  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
Clube                                                               
Avaí - SC                  57.5           -27.8              95.5   
América - MG               52.6           -17.4              78.0   
Figueirense - SC           54.8           -19.2              96.0   
Coritiba - PR              52.4           -12.2              97.0   
Chapecoense - SC           51.7           -14.4              80.9   

                  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                    
Avaí - SC                       5.5                30.0  
América - MG                    5.8                34.4  
Figueirense - SC                8.0                34.0  
Coritiba - PR                   5.9                36.6  
Chapecoense - SC                4.0                35.7

In [65]:
media_de_equipe_da_estatistica("Derrotas")

<h2>27 - Identificar OS TIMES que mais vezes ficaram no G4.</h2>

In [66]:
todas_as_equipes_do_g4 = df[df["Colocação"] < 5]
todas_as_equipes_do_g4 = todas_as_equipes_do_g4['Clube'].value_counts()

In [67]:
fig = px.bar(todas_as_equipes_do_g4, title="Times que mais vezes terminaram no G4")
fig.update_layout(
    showlegend=False,
    yaxis_title="Qtd. de Vezes",
    xaxis_title = "",
)
fig.update_traces(
    hovertemplate="Clube: %{x}<br>Frequência: %{y}<extra></extra>"
)
fig.show()

<h2>28 - Identificar OS TIMES com as melhores médias de colocação.</h2>

In [68]:
media_com_filtro.sort_values('Colocação').head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                   
Atlético Mineiro - MG        5.2    63.2      17.8      9.8      10.4   
Grêmio - RS                  5.5    61.9      17.4      9.8      10.8   
Palmeiras - SP               5.6    63.5      18.2      9.0      10.8   
Flamengo - RJ                6.1    62.9      18.0      9.2      10.8   
Corinthians - SP             7.0    58.9      15.7     11.9      10.4   

                       Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                               
Atlético Mineiro - MG         55.8           41.8            14.0   
Grêmio - RS                   50.4           37.9            12.5   
Palmeiras - SP                56.4           39.9            16.5   
Flamengo - RJ                 56.0           41.3            14.7   
Corinthians - SP              45.7           35.8             9.9   

                       Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                           
Atlético Mineiro - MG              88.8                5.0                55.1  
Grêmio - RS                        88.5                4.7                53.8  
Palmeiras - SP                     94.1                5.5                55.3  
Flamengo - RJ                      79.2                4.8                54.7  
Corinthians - SP                   76.3                3.7                51.3

In [69]:
media_de_equipe_da_estatistica('Colocação')

<h2>29 - Identificar OS TIMES com as melhores médias de aproveitamento.</h2>

In [70]:
media_com_filtro.sort_values('Aproveitamento (%)', ascending=False).head()

Colocação  Pontos  Vitórias  Empates  Derrotas  \
Clube                                                                   
Palmeiras - SP               5.6    63.5      18.2      9.0      10.8   
Atlético Mineiro - MG        5.2    63.2      17.8      9.8      10.4   
Flamengo - RJ                6.1    62.9      18.0      9.2      10.8   
Grêmio - RS                  5.5    61.9      17.4      9.8      10.8   
Corinthians - SP             7.0    58.9      15.7     11.9      10.4   

                       Gols Feitos  Gols Sofridos  Saldos de Gols  \
Clube                                                               
Palmeiras - SP                56.4           39.9            16.5   
Atlético Mineiro - MG         55.8           41.8            14.0   
Flamengo - RJ                 56.0           41.3            14.7   
Grêmio - RS                   50.4           37.9            12.5   
Corinthians - SP              45.7           35.8             9.9   

                       Cartões Amarelos  Cartões Vermelhos  Aproveitamento (%)  
Clube                                                                           
Palmeiras - SP                     94.1                5.5                55.3  
Atlético Mineiro - MG              88.8                5.0                55.1  
Flamengo - RJ                      79.2                4.8                54.7  
Grêmio - RS                        88.5                4.7                53.8  
Corinthians - SP                   76.3                3.7                51.3

In [71]:
media_de_equipe_da_estatistica('Aproveitamento (%)')

<h2>30 - Agrupar os times pelo TOTAL de cada estado</h2>

In [72]:
df_com_estados = df.copy()
df_com_estados['Estado'] = df['Clube'].apply(lambda x: x[-2:])
df_com_estados

Ano  Colocação                  Clube  Pontos  Vitórias  Empates  \
0    2012          1        Fluminense - RJ      77        22       11   
1    2012          2  Atlético Mineiro - MG      72        20       12   
2    2012          3            Grêmio - RS      71        20       11   
3    2012          4         São Paulo - SP      66        20        6   
4    2012          5     Vasco da Gama - RJ      58        16       10   
..    ...        ...                    ...     ...       ...      ...   
235  2023         16             Bahia - BA      44        12        8   
236  2023         17            Santos - SP      43        11       10   
237  2023         18             Goiás - GO      38         9       11   
238  2023         19          Coritiba - PR      30         8        6   
239  2023         20           América - MG      24         5        9   

     Derrotas  Gols Feitos  Gols Sofridos  Saldos de Gols  Cartões Amarelos  \
0           5           61             33              28                85   
1           6           64             37              27               104   
2           7           56             33              23               102   
3          12           59             37              22                93   
4          12           45             44               1                91   
..        ...          ...            ...             ...               ...   
235        18           50             53              -3                96   
236        17           39             64             -25               103   
237        18           36             53             -17               108   
238        24           41             73             -32                68   
239        24           42             81             -39                86   

     Cartões Vermelhos  Aproveitamento (%) Estado  
0                    3                  67     RJ  
1                   10                  63     MG  
2                    9                  62     RS  
3                    6                  57     SP  
4                    3                  50     RJ  
..                 ...                 ...    ...  
235                  8                  38     BA  
236                  9                  37     SP  
237                  8                  33     GO  
238                  6                  26     PR  
239                 11                  21     MG  

[240 rows x 14 columns]

In [73]:
soma_infos_por_estados = df_com_estados.groupby('Estado').sum().drop(columns=['Clube', 'Ano', 'Colocação', 'Aproveitamento (%)'])
soma_infos_por_estados.index.name = 'Estado'
soma_infos_por_estados

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Estado                                                                    
AL          32         8        8        22           24             58   
BA         632       162      146       224          611            705   
CE         483       124      111       145          414            430   
GO         522       132      126       198          475            627   
MG        1464       402      258       328         1272           1126   
MT         139        34       37        43          105            118   
PE         463       123       97       198          424            596   
PR        1014       272      198       328          897            984   
RJ        2295       631      406       559         2004           1884   
RS        1393       380      253       279         1145            962   
SC         693       169      186       329          625            978   
SP        3308       894      630       680         2811           2345   

        Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  
Estado                                                       
AL                 -34                90                  4  
BA                 -94              1275                 77  
CE                 -16               934                 72  
GO                -152              1131                 70  
MG                 146              2272                137  
MT                 -13               268                 22  
PE                -172               919                 54  
PR                 -87              1924                105  
RJ                 120              3707                237  
RS                 183              2270                134  
SC                -353              1640                101  
SP                 466              5129                303

<h2>31 - Agrupar os times pela MÉDIA de cada estado</h2>

In [74]:
media_infos_por_estados = df_com_estados.groupby('Estado').mean(numeric_only=True).drop(columns=['Ano', 'Colocação']).round(1)
media_infos_por_estados.index.name = "Estado"
media_infos_por_estados

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Estado                                                                    
AL        32.0       8.0      8.0      22.0         24.0           58.0   
BA        45.1      11.6     10.4      16.0         43.6           50.4   
CE        48.3      12.4     11.1      14.5         41.4           43.0   
GO        43.5      11.0     10.5      16.5         39.6           52.2   
MG        56.3      15.5      9.9      12.6         48.9           43.3   
MT        46.3      11.3     12.3      14.3         35.0           39.3   
PE        42.1      11.2      8.8      18.0         38.5           54.2   
PR        48.3      13.0      9.4      15.6         42.7           46.9   
RJ        54.6      15.0      9.7      13.3         47.7           44.9   
RS        58.0      15.8     10.5      11.6         47.7           40.1   
SC        38.5       9.4     10.3      18.3         34.7           54.3   
SP        57.0      15.4     10.9      11.7         48.5           40.4   

        Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  \
Estado                                                        
AL               -34.0              90.0                4.0   
BA                -6.7              91.1                5.5   
CE                -1.6              93.4                7.2   
GO               -12.7              94.2                5.8   
MG                 5.6              87.4                5.3   
MT                -4.3              89.3                7.3   
PE               -15.6              83.5                4.9   
PR                -4.1              91.6                5.0   
RJ                 2.9              88.3                5.6   
RS                 7.6              94.6                5.6   
SC               -19.6              91.1                5.6   
SP                 8.0              88.4                5.2   

        Aproveitamento (%)  
Estado                      
AL                    28.0  
BA                    39.1  
CE                    41.8  
GO                    37.7  
MG                    49.0  
MT                    40.0  
PE                    36.5  
PR                    42.0  
RJ                    47.4  
RS                    50.5  
SC                    33.4  
SP                    49.6

<h2>32 - Agrupar os times pelo TOTAL de CADA REGIÃO</h2>

In [75]:
regioes = {
    "Norte": ["AC", "AM", "AP", "PA", "RO", "RR", "TO"],
    "Nordeste": ["AL", "BA", "CE", "MA", "PB", "PE", "PI", "RN", "SE"],
    "Centro-Oeste": ["DF", "GO", "MS", "MT"],
    "Sudeste": ["SP", "RJ", "ES", "MG"],
    "Sul": ["RS", "PR", "SC"]
}

In [76]:
def atribuir_regiao(x):
    for regiao in regioes.keys():
        if x in regioes[regiao]:
            return regiao

In [77]:
df_com_estados['Região'] = df_com_estados['Estado'].apply(lambda x: atribuir_regiao(x))

In [78]:
soma_infos_por_regiao = df_com_estados.groupby('Região').sum().drop(columns=['Clube', 'Ano', 'Colocação', 'Aproveitamento (%)', 'Estado'])
soma_infos_por_regiao

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Região                                                                          
Centro-Oeste     661       166      163       241          580            745   
Nordeste        1610       417      362       589         1473           1789   
Sudeste         7067      1927     1294      1567         6087           5355   
Sul             3100       821      637       936         2667           2924   

              Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  
Região                                                             
Centro-Oeste            -165              1399                 92  
Nordeste                -316              3218                207  
Sudeste                  732             11108                677  
Sul                     -257              5834                340

<h2>33 - Agrupar os times pela MÉDIA de CADA REGIÃO</h2>

In [79]:
media_infos_por_regiao = df_com_estados.groupby('Região').mean(numeric_only=True).drop(columns=['Ano', 'Colocação']).round(1)
media_infos_por_regiao

Pontos  Vitórias  Empates  Derrotas  Gols Feitos  Gols Sofridos  \
Região                                                                          
Centro-Oeste    44.1      11.1     10.9      16.1         38.7           49.7   
Nordeste        44.7      11.6     10.1      16.4         40.9           49.7   
Sudeste         56.1      15.3     10.3      12.4         48.3           42.5   
Sul             49.2      13.0     10.1      14.9         42.3           46.4   

              Saldos de Gols  Cartões Amarelos  Cartões Vermelhos  \
Região                                                              
Centro-Oeste           -11.0              93.3                6.1   
Nordeste                -8.8              89.4                5.8   
Sudeste                  5.8              88.2                5.4   
Sul                     -4.1              92.6                5.4   

              Aproveitamento (%)  
Região                            
Centro-Oeste                38.1  
Nordeste                    38.8  
Sudeste                     48.7  
Sul                         42.8